File: counter.ts
import { Account, Pubkey, Result, i64, u8, Signer } from "@turbin3/poseidon";

export default class VoteProgram {
    static PROGRAM_ID = new Pubkey("HC2oqz2p6DEWfrahenqdq2moUcga9c9biqRBcdK3XKU1");

    public initialize(state: VoteState, hash: Uint8Array, user: Signer): Result {
        state.derive(["vote", hash])
            .init()
        state.vote = new i64(0)
    }

    public upvote(state: VoteState, hash: Uint8Array): Result {
        state.derive(["vote", hash])
        state.vote = state.vote.add(1)
    }

    public downvote(state: VoteState, hash: Uint8Array): Result {
        state.derive(["vote", hash])
        state.vote = state.vote.sub(1)
    }
}

export interface VoteState extends Account {
    vote: i64
    // bump: u8
}

File: counter.rs
use anchor::prelude::*;
#[program]
pub mod VoteProgram {
    declare_id!("HC2oqz2p6DEWfrahenqdq2moUcga9c9biqRBcdK3XKU1");
    pub fn initialize(ctx: Context<InitializeContext>, hash: Vec<u8>) -> Result<()> {}
    pub fn upvote(ctx: Context<UpvoteContext>, hash: Vec<u8>) -> Result<()> {}
    pub fn downvote(ctx: Context<DownvoteContext>, hash: Vec<u8>) -> Result<()> {}
}
#[derive(Accounts)]
pub struct InitializeContext<'info> {
    #[account()]
    pub user: Signer<'info>,
    # [account (init , payer = user , bump seeds = [b "vote" , hash])]
    pub state: Account<'info, VoteState>,
    pub system_program: Program<'info, System>,
}
#[derive(Accounts)]
pub struct UpvoteContext<'info> {
    # [account (seeds = [b "vote" , hash])]
    pub state: Account<'info, VoteState>,
}
#[derive(Accounts)]
pub struct DownvoteContext<'info> {
    # [account (seeds = [b "vote" , hash])]
    pub state: Account<'info, VoteState>,
}
#[account]
pub struct VoteState {
    pub vote: i64,
}
